File: prepare.py
import os
import json
from PIL import Image
import glob
from pathlib import Path

def convert_to_webp(input_path, output_path, size=(512, 512)):
    """Convert image to WebP format with specified size"""
    try:
        with Image.open(input_path) as img:
            # Convert to RGBA if necessary
            if img.mode != 'RGBA':
                img = img.convert('RGBA')
            
            # Resize image maintaining aspect ratio
            img.thumbnail(size, Image.Resampling.LANCZOS)
            
            # Create a new image with padding if necessary
            new_img = Image.new('RGBA', size, (0, 0, 0, 0))
            offset = ((size[0] - img.size[0]) // 2, (size[1] - img.size[1]) // 2)
            new_img.paste(img, offset)
            
            # Save as WebP
            new_img.save(output_path, 'WEBP', quality=90)
            return True
    except Exception as e:
        print(f"Error converting {input_path}: {str(e)}")
        return False

def process_images(input_dir, output_dir):
    """Process all images in input directory and convert to WebP"""
    # Create output directory if it doesn't exist
    Path(output_dir).mkdir(parents=True, exist_ok=True)
    
    # Supported input formats
    input_formats = ['*.png', '*.jpg', '*.jpeg']
    processed_files = []
    
    # Process each image
    for format in input_formats:
        for input_path in glob.glob(os.path.join(input_dir, format)):
            filename = os.path.basename(input_path)
            name_without_ext = os.path.splitext(filename)[0]
            output_path = os.path.join(output_dir, f"{name_without_ext}.webp")
            
            if convert_to_webp(input_path, output_path):
                processed_files.append({
                    "original": filename,
                    "webp": f"{name_without_ext}.webp",
                    "emoji": "😀"  # Default emoji
                })
    
    return processed_files

def generate_config(processed_files, output_dir):
    """Generate configuration file"""
    config = {
        "title": "My Awesome Stickers",
        "short_name": "my_awesome_stickers",
        "stickers": processed_files
    }
    
    # Save config file
    config_path = os.path.join(output_dir, "sticker_config.json")
    with open(config_path, 'w', encoding='utf-8') as f:
        json.dump(config, f, indent=4, ensure_ascii=False)
    
    return config_path

def main():
    # Setup directories
    input_dir = "input"
    output_dir = "output"
    
    # Create input directory if it doesn't exist
    Path(input_dir).mkdir(parents=True, exist_ok=True)
    
    print("Processing images...")
    processed_files = process_images(input_dir, output_dir)
    
    if not processed_files:
        print("No images found in input directory!")
        return
    
    print(f"Processed {len(processed_files)} images")
    
    # Generate configuration file
    config_path = generate_config(processed_files, output_dir)
    print(f"\nConfiguration file generated: {config_path}")
    print("\nYou can now edit the configuration file to:")
    print("1. Change the sticker pack title and short name")
    print("2. Modify emojis for each sticker")
    print("3. Add or remove stickers from the pack")
    print("\nAfter editing, run main.py to create the sticker pack")

if __name__ == "__main__":
    main() 

File: prepare.go
package main

import (
	"encoding/json"
	"fmt"
	"image"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"github.com/chai2010/webp"
	"github.com/disintegration/imaging"
)

type Sticker struct {
	Original string `json:"original"`
	Webp     string `json:"webp"`
	Emoji    string `json:"emoji"`
}

type Config struct {
	Title     string    `json:"title"`
	ShortName string    `json:"short_name"`
	Stickers  []Sticker `json:"stickers"`
}

func convertToWebp(inputPath, outputPath string, size image.Point) error {
	srcImg, err := imaging.Open(inputPath)
	if err != nil {
		return fmt.Errorf("open %s: %w", inputPath, err)
	}

	// Convert to RGBA
	rgba := imaging.ToRGBA(srcImg)

	// Resize maintaining aspect ratio
	resized := imaging.Fit(rgba, size.X, size.Y, imaging.Lanczos)

	// Pad to exact size with transparent background
	dst := image.NewRGBA(size)
	offset := imaging.Center(size.X, size.Y, resized.Bounds().Dx(), resized.Bounds().Dy())
	drawer := imaging.NewDrawer(dst)
	drawer.Draw(offset.X, offset.Y, resized)

	// Save as WebP
	opts := &webp.Options{Lossless: false, Quality: 90}
	if err := webp.EncodeFile(outputPath, dst, opts); err != nil {
		return fmt.Errorf("encode webp %s: %w", outputPath, err)
	}
	return nil
}

func processImages(inputDir, outputDir string) ([]Sticker, error) {
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return nil, err
	}

	var processed []Sticker
	err := filepath.WalkDir(inputDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		ext := strings.ToLower(filepath.Ext(path))
		if ext != ".png" && ext != ".jpg" && ext != ".jpeg" {
			return nil
		}

		filename := d.Name()
		nameWithoutExt := strings.TrimSuffix(filename, filepath.Ext(filename))
		webpPath := filepath.Join(outputDir, nameWithoutExt+".webp")

		if err := convertToWebp(path, webpPath, image.Pt(512, 512)); err != nil {
			fmt.Printf("Error converting %s: %v\n", filename, err)
			return nil
		}

		processed = append(processed, Sticker{
			Original: filename,
			Webp:     nameWithoutExt + ".webp",
			Emoji:    "😀", // Default
		})
		fmt.Printf("Processed %s -> %s\n", filename, nameWithoutExt+".webp")
		return nil
	})
	return processed, err
}

func main() {
	inputDir := "input"
	outputDir := "output"

	if err := os.MkdirAll(inputDir, 0755); err != nil {
		fmt.Printf("Error creating input dir: %v\n", err)
		return
	}

	fmt.Println("Processing images...")
	stickers, err := processImages(inputDir, outputDir)
	if err != nil {
		fmt.Printf("Error processing images: %v\n", err)
		return
	}
	if len(stickers) == 0 {
		fmt.Println("No images found in input directory!")
		return
	}

	config := Config{
		Title:     "My Awesome Stickers",
		ShortName: "my_awesome_stickers",
		Stickers:  stickers,
	}

	configPath := filepath.Join(outputDir, "sticker_config.json")
	data, err := json.MarshalIndent(config, "", "    ")
	if err != nil {
		fmt.Printf("Error marshaling config: %v\n", err)
		return
	}

	if err := os.WriteFile(configPath, data, 0644); err != nil {
		fmt.Printf("Error writing config: %v\n", err)
		return
	}

	fmt.Printf("Processed %d images\n", len(stickers))
	fmt.Printf("Configuration generated: %s\n", configPath)
	fmt.Println("\nEdit sticker_config.json to customize:")
	fmt.Println("1. Change title and short_name")
	fmt.Println("2. Set custom emojis for stickers")
	fmt.Println("3. Add/remove stickers")
	fmt.Println("\nThen run: go run main.go")
}
